Домашнее задание за 17 апреля!

ITevgeniydav2022 opened the pull request:
Так увлеклись разработкой, что сделали больше необходимого чем требуется к сдаче первой домашки. Надеемся это не критично.

ildar1902 committed:
внёс незначительные изменения в pom.xml, чтобы сделать первый комит

ITevgeniydav2022 committed:
Добавлены классы Dto согласно спецификации: AdsComment, Ads, CreateAds, CreateUser, FullAds, NewPassword, ResponseWrapper, User. Добавлен класс Controller согласно спецификации: User.

ITevgeniydav2022 committed:
Merge remote-tracking branch 'origin/dev' into dev

ITevgeniydav2022 committed:
Добавлены классы Dto согласно спецификации: ResponseWrapperAds, переименован класс ResponseWrapperComment.

ITevgeniydav2022 committed:
Merge branch 'evgeny-d' into dev

KatOli4ka committed:
Сделала полностью 1 дз и практически часть 2 дз. Очень много описывать. Осталось сделать мапперы, логгер. Добавила несколько зависимостей в помник. Пока все сырое - будет позже доработка после коментов наставника

ITevgeniydav2022 committed:
Подключил Базу данных.

KatOli4ka committed:
1.добавила в помник библиотеку маппера
2.доработала классы с wrapper-ом.
3.исправила у контроллерах оббъявления и комментария ссылку на класс wrapper.
4.ждемс сами мапперы........

KatOli4ka committed:
Merge remote-tracking branch 'origin/dev' into dev

ildar1902 committed:
сделал мапперы

KatOli4ka committed:
1. исправила мапперы, есть вопросы к наставнику по аннотациям
2. создала исключения
3. подредачила сервисы и контроллеры исходя из новых методов и классов (есть закоменнтированные недоработанные методы. думаю в 3 дз появятся подсказки к ним)
4.добавила в вэбсекьюрити проверку пароля на правильность.
ЕЩЕ РАЗ! Все будет еще 300 раз переделываться - все сырое я думаю.
МОЖНО ДОБАВЛЯТЬ ЛОГГЕР!

KatOli4ka committed:
подправила метод сохранения фото в бд

TKaterinaS committed:
добавила лог файл в exml,добавила логи в сервисах и в Config и в контроллер автоматизации

dmitry-bizin commented on pom.xml:
```diff
@@ -65,6 +93,23 @@
+                <artifactId>maven-compiler-plugin</artifactId>
+                <version>3.10.1</version>
+                <configuration>
+                    <source>11</source> <!-- depending on your project -->
```
а почему 11 версия? у вас выше 17 стоит, а здесь 11. Какая в итоге? Должна быть одинаковая

dmitry-bizin commented on pom.xml:
```diff
@@ -15,40 +15,68 @@
+        <dependency>
+            <groupId>org.postgresql</groupId>
+            <artifactId>postgresql</artifactId>
+            <version>42.5.4</version>
```
версию лучше не писать

dmitry-bizin commented on src/main/java/com/example/graduate_work_team2/checkConfig/WebSecurityConfig.java:
```diff
@@ -0,0 +1,66 @@
+                    )
+                    .cors().disable()
+                    .httpBasic(withDefaults());
+        } catch (Exception e) {
```
а зачем здесь try/catch? это не нужно

dmitry-bizin commented on src/main/java/com/example/graduate_work_team2/checkConfig/WebSecurityConfig.java:
```diff
@@ -0,0 +1,66 @@
+        }
+        try {
+            return http.build();
+        } catch (Exception e) {
```
а зачем здесь try/catch? это не нужно

dmitry-bizin commented on src/main/java/com/example/graduate_work_team2/checkConfig/WebSecurityConfig.java:
```diff
@@ -0,0 +1,66 @@
+                                    .mvcMatchers(AUTH_WHITELIST).permitAll()
+                                    .mvcMatchers("/ads/**", "/users/**").authenticated()
+                    )
+                    .cors().disable()
```
вместо diasble() должно быть and()

dmitry-bizin commented on src/main/java/com/example/graduate_work_team2/controller/AdsController.java:
```diff
@@ -0,0 +1,177 @@
+    )
+    @GetMapping("/{adsId}")
+    public ResponseEntity<FullAdsDto> getFullAdsDto(@PathVariable("adsId") Long adsId) {
+        return ResponseEntity.ok(adsMapper.toFullAdsDto(adsService.getAdsById(adsId)));
```
мапперы должны использоваться внутри сервисов. Сервисы принимают и возвращают dto, внутри работают с сущностями. Сейчас сущности проскальзывают в контроллер - это не очень хорошо в плане транзакционного контекста

dmitry-bizin commented on src/main/java/com/example/graduate_work_team2/controller/RegistrationController.java:
```diff
@@ -0,0 +1,50 @@
+    )
+    @PostMapping("/register")
+    public ResponseEntity<Void> register(@Valid @RequestBody RegisterReqDto req) {
+       registrationService.register(userMapper.toEntity(req));
```
мапперы должны использоваться внутри сервисов. Сервисы принимают и возвращают dto, внутри работают с сущностями. Сейчас сущности проскальзывают в контроллер - это не очень хорошо в плане транзакционного контекста

dmitry-bizin commented on src/main/java/com/example/graduate_work_team2/controller/UserController.java:
```diff
@@ -0,0 +1,106 @@
+            }
+    )
+    @PatchMapping(value = "/me/image", consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
+    public ResponseEntity<String> updateUserImage(@RequestBody MultipartFile image, Authentication authentication) throws IOException {
```
аннотация неверная - так не будет работать

BizinMitya commented on src/main/java/com/example/graduate_work_team2/dto/UserDto.java:
```diff
@@ -0,0 +1,39 @@
+    private String lastName;
+    private String phone;
+    @JsonIgnore
+    private String regDate = String.valueOf(LocalDateTime.now());
```
а зачем строка? почему бы не хранить в LocalDateTime? со строкой неудобно работать

BizinMitya commented on src/main/java/com/example/graduate_work_team2/mapper/AdsMapper.java:
```diff
@@ -0,0 +1,43 @@
+@Mapper
+public interface AdsMapper {
+
+//    @Mapping(target = "author", source = "author.id") - в дальнейшем уточнить!
```
а что уточнить?

dmitry-bizin replied:
без этой аннотации маппинга не будет, потому что как библиотека поймёт, что надо взять именно id и положить именно в author? Никак. Надо ей сказать. Ниже аналогично.

dmitry-bizin commented on src/main/java/com/example/graduate_work_team2/mapper/AdsMapper.java:
```diff
@@ -0,0 +1,43 @@
+
+    @Mapping(target = "authorFirstName", source = "author.firstName")
+    @Mapping(target = "authorLastName", source = "author.lastName")
+//    @Mapping(target = "phone", source = "author.phone") - в дальнейшем уточнить!
```
здесь можно не писать, потому что названия полей и типы совпадают

dmitry-bizin commented on src/main/java/com/example/graduate_work_team2/mapper/AdsMapper.java:
```diff
@@ -0,0 +1,43 @@
+    @Mapping(target = "authorFirstName", source = "author.firstName")
+    @Mapping(target = "authorLastName", source = "author.lastName")
+//    @Mapping(target = "phone", source = "author.phone") - в дальнейшем уточнить!
+//    @Mapping(target = "email", source = "author.email") - аналогично
```
здесь можно не писать, потому что названия полей и типы совпадают

dmitry-bizin commented on src/main/java/com/example/graduate_work_team2/mapper/CommentMapper.java:
```diff
@@ -0,0 +1,32 @@
+    @Mapping(target = "createdAt", ignore = true)
+    CommentDto toCommentDto(Comment entity);
+
+//    @Mapping(target = "author", source = "author.id") -уточнить!!!!!
```
надо

dmitry-bizin commented on src/main/java/com/example/graduate_work_team2/mapper/CommentMapper.java:
```diff
@@ -0,0 +1,32 @@
+
+    List<CommentDto> toCommentDto(Collection<Comment> entity);
+
+//    List<Comment> fromCommentDto(Collection<CommentDto> dto); - в дальнейшем уточнить!
```
а здесь что уточнить? если нужен этот маппинг, то надо оставить, не нужно - не надо)

dmitry-bizin commented on src/main/java/com/example/graduate_work_team2/mapper/AdsMapper.java:
```diff
@@ -0,0 +1,43 @@
+    FullAdsDto toFullAdsDto(Ads entity);
+    List<AdsDto> toAdsDto(Collection<Ads> entity);
+
+//    List<Ads> fromAdsDto(Collection<AdsDto> dto); - аналгично
```
а здесь что? если нужен этот маппинг, то надо оставить, не нужно - не надо)

dmitry-bizin commented on src/main/java/com/example/graduate_work_team2/mapper/UserMapper.java:
```diff
@@ -0,0 +1,42 @@
+
+    List<UserDto> toUserDto(Collection<User> entity);
+
+//    List<User> fromUserDto(Collection<UserDto> dto);- в дальнейшем уточнить!
```
а здесь что? если нужен этот маппинг, то надо оставить, не нужно - не надо)

dmitry-bizin commented on src/main/java/com/example/graduate_work_team2/service/impl/ImageServiceImpl.java:
```diff
@@ -0,0 +1,101 @@
+            Path filePath = Path.of(updatedImage.getFilePath());
+            Files.deleteIfExists(filePath);
+            try (
+                    InputStream is = imageFile.getInputStream();
```
Files.write - в одну строку можно записать в файл

KatOli4ka committed:
1.поменяла версию мавена с 11 на 17 как сам проект.
2.убрала версию постгреса в помпнике
3.исправила многие методы в многих классах
4.добавила плагины в помник
5.подредачила все мапперы

KatOli4ka committed:
1.исправила остальные ошибки в написанных классах
2.добавила логи в методы

KatOli4ka committed:
1.вернула конструкторы в юзердто, кот.были в тз изначально
2. исправила ошибку в методе создания юзера

TatyanaBogodvid committed:
Добавлен java-doc

KatOli4ka committed:
1.убрала лишний джавадок в контроллерах
2.добавила коды ответов в контроллеры
3.добавила дто в соответствии с новой спецификацией.
4.убрала лишнюю инфу исходя из новой спец-ии.

KatOli4ka committed:
1. работает только регистрация и авторизация.
2. мапперы рукописные - с аннотациями так и не разобралась
3. при любом другом действии в личном кабинете - вызывается метод получения всех объявлений пользователя и куча серверных ошибок. Пока не пойму - почему так......
4.Пока всё в процессе дальнейшей разработки.....

ITevgeniydav2022 committed:
Удалил пакет checkConfig с лишним классом WebSecurityConfig.java

KatOli4ka committed:
проверка

KatOli4ka committed:
Merge remote-tracking branch 'origin/dev' into dev

KatOli4ka committed:
проверка 2

KatOli4ka committed:
проверка 1000

KatOli4ka committed:
Merge branch 'kate-o' into dev

# Conflicts:
#	src/main/java/com/example/graduate_work_team2/repository/AdsRepository.java

KatOli4ka committed:
1. передела все классы
2. пока работает авторизация и регистрация

ITevgeniydav2022 committed:
Сделал проект рабочим)

dmitry-bizin commented on pom.xml:
```diff
@@ -15,41 +15,87 @@
     <properties>
         <java.version>17</java.version>
+        <org.mapstruct.version>1.5.4.Final</org.mapstruct.version>
+        <maven.compiler.source>1.8</maven.compiler.source>
```
почему версия здесь 8, когда выше стоит 17? должна быть везде одна  и та же версия

dmitry-bizin commented on src/main/java/com/example/graduate_work_team2/configuration/BasicAuthCorsFilter.java:
```diff
@@ -9,8 +11,10 @@
+/**Базовый фильтр проверки подлинности**/
 @Component
 public class BasicAuthCorsFilter extends OncePerRequestFilter {
+    private static final Logger logger = LoggerFactory.getLogger(BasicAuthCorsFilter.class);
```
LOGGER, т.к. static И final

dmitry-bizin commented on src/main/java/com/example/graduate_work_team2/configuration/WebSecurityConfig.java:
```diff
@@ -0,0 +1,96 @@
+            "/webjars/**",
+            "/login", "/register",
+            "/ads", "/ads/*/image",
+            "/users", "/users/*/image"
```
/users здесь быть не должен, такого эндпоинта и нет в спецификации

dmitry-bizin commented on src/main/java/com/example/graduate_work_team2/configuration/WebSecurityConfig.java:
```diff
@@ -0,0 +1,96 @@
+            "/ads", "/ads/*/image",
+            "/users", "/users/*/image"
+    };
+    private final MyUserDetailsService myUserDetailsService;
```
UserDetailsService - инжектить надо по интерфейсу всегда

dmitry-bizin commented on src/main/java/com/example/graduate_work_team2/configuration/WebSecurityConfig.java:
```diff
@@ -0,0 +1,96 @@
+                .csrf().disable()
+                .authorizeHttpRequests((authz) ->
+                        authz
+                                .mvcMatchers(HttpMethod.GET, "/ads").permitAll()
```
это не нужно, у вас уже есть в AUTH_WHITELIST

dmitry-bizin commented on src/main/java/com/example/graduate_work_team2/configuration/WebSecurityConfig.java:
```diff
@@ -0,0 +1,96 @@
+                .authorizeHttpRequests((authz) ->
+                        authz
+                                .mvcMatchers(HttpMethod.GET, "/ads").permitAll()
+                                .mvcMatchers(HttpMethod.GET, "/ads/images/**").permitAll()
```
это не нужно, у вас уже есть в AUTH_WHITELIST

dmitry-bizin commented on src/main/java/com/example/graduate_work_team2/controller/AdsController.java:
```diff
@@ -0,0 +1,182 @@
+    )
+
+    @PostMapping(consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
+    public ResponseEntity<AdsDto> addAds(@RequestPart("dto") CreateAdsDto dto,
```
здесь название части должно называться properties, а не dto - см. спецификацию, я даже писал замечания вроде бы в первой домашке и вы исправляли, а сейчас снова сломали. Поэтому и не работает у вас добавление объявления

dmitry-bizin commented on src/main/java/com/example/graduate_work_team2/controller/CommentController.java:
```diff
@@ -0,0 +1,127 @@
+        log.info("Был вызван метод контроллера для изменения комментария объявления");
+        CommentDto updatedCommentDto = commentService.updateComment(adsId, comId,
+                updateCommentDto, authentication);
+        if (updateCommentDto.equals(updatedCommentDto)) {
```
не очень понял логику этого. А если я пошлю запрос на обновление и ничего не поменяю в данных? у вас будет ошибка, хотя не должна быть. 401 здесь вообще возвращать не надо, т.к. метод под авторизацией и если он начал выполнение, значит запрос авторизованный и спринг проверил его уже, самим проверять это дополнительно не нужно

dmitry-bizin commented on src/main/java/com/example/graduate_work_team2/controller/CommentController.java:
```diff
@@ -0,0 +1,127 @@
+        if (commentService.deleteCommentDto(adsId, comId)) {
+            return ResponseEntity.ok().build();
+        }
+        return ResponseEntity.status(HttpServletResponse.SC_UNAUTHORIZED).build();
```
401 возвращать не надо, т.к. метод под авторизацией и если он начал выполнение, значит запрос авторизованный и спринг проверил его уже, самим возвращать это не нужно.

Здесь надо возвращать 403, т.к. попытка удалить тем, кому это не разрешено

dmitry-bizin commented on src/main/java/com/example/graduate_work_team2/service/impl/AdsServiceImpl.java:
```diff
@@ -0,0 +1,125 @@
+        log.info("Был вызван метод получения объявлений авторизованного пользователя");
+        User user = userService.findAuthorizationUser().orElseThrow(UserNotFoundException::new);
+        Collection<Ads> adsList = adsRepository.findAll();
+        Collection<Ads> user_sAds = adsList.stream().filter(x -> x.getAuthor().equals(user)).toList();
```
camelCase, а не snake_case

dmitry-bizin replied:
много раз говорил, что вытягивать все строки из таблицы, а затем фильтровать их в коде - очень-очень-очень плохая практика, так вы рано или поздно нарвётесь на OutOfMemoryError или просто на тормоза приложения на пусто месте. Надо фильтровать на уровне БД запросами, а не в коде

dmitry-bizin commented on src/main/java/com/example/graduate_work_team2/service/impl/AdsServiceImpl.java:
```diff
@@ -0,0 +1,125 @@
+            return true;
+        }
+        log.info("Удаление прошло успешно!");
+        throw new UserNotFoundException();
```
здесь надо возвращать 403, т.к. попытка удалить тем, кому это не разрешено

dmitry-bizin commented on src/main/java/com/example/graduate_work_team2/service/impl/AdsServiceImpl.java:
```diff
@@ -0,0 +1,125 @@
+            return adsMapper.toDto(adsRepository.save(updatedAds));
+        }
+        log.info("Измененное объявление добавлено!");
+        throw new UserNotFoundException();
```
здесь надо возвращать 403, т.к. попытка обновить тем, кому это не разрешено

dmitry-bizin commented on src/main/java/com/example/graduate_work_team2/service/impl/CommentServiceImpl.java:
```diff
@@ -0,0 +1,103 @@
+            commentRepository.save(updatedComment);
+            return commentMapper.toDto(updatedComment);
+        }
+        return updateComment;
```
здесь надо возвращать 403, т.к. попытка обновить тем, кому это не разрешено

dmitry-bizin commented on pom.xml:
```diff
@@ -15,41 +15,87 @@
+
     <dependencies>
+        <dependency>
+            <groupId>org.mapstruct</groupId>
```
mapstruct-ские зависимости удалите, потому что вы не используете mapstruct. Из-за этого у вас у некоторых не запускается проект как раз. Надо удалить зависимости, удалить папку target и пересобрать проект мавеном

dmitry-bizin commented on pom.xml:
```diff
@@ -65,6 +111,38 @@
+                        <path>
+                            <groupId>org.mapstruct</groupId>
+                            <artifactId>mapstruct-processor</artifactId>
+                            <version>${org.mapstruct.version}</version>
```
mapstruct-ские зависимости удалите, потому что вы не используете mapstruct. Из-за этого у вас у некоторых не запускается проект как раз. Надо удалить зависимости, удалить папку target и пересобрать проект мавеном

dmitry-bizin commented on pom.xml:
```diff
@@ -65,6 +111,38 @@
+                    </path>
+                        <path>
+                            <groupId>org.projectlombok</groupId>
+                            <artifactId>lombok-mapstruct-binding</artifactId>
```
mapstruct-ские зависимости удалите, потому что вы не используете mapstruct. Из-за этого у вас у некоторых не запускается проект как раз. Надо удалить зависимости, удалить папку target и пересобрать проект мавеном

ITevgeniydav2022 committed:
Проведён рефакторинг кода. Написаны контроллеры изображений.

dmitry-bizin commented on src/main/java/com/example/graduate_work_team2/configuration/WebSecurityConfig.java:
```diff
@@ -79,7 +79,7 @@ public SecurityFilterChain filterChain(HttpSecurity http)  throws Exception{
                                 .mvcMatchers(HttpMethod.GET, "/ads/images/**").permitAll()
                                 .mvcMatchers(AUTH_WHITELIST).permitAll()
-                                .mvcMatchers("/ads/**", "/users/**").authenticated()
+                                .mvcMatchers("/ads/**", "/users/**").permitAll()
```
о как, а почему без авторизации доступны эндпоинты?) это неверно, надо здесь чтобы было authenticated()

dmitry-bizin commented on src/main/java/com/example/graduate_work_team2/controller/CommentController.java:
```diff
@@ -44,16 +45,17 @@ public class CommentController {
     )
 
-    @GetMapping("/{id}/comments")
+    @GetMapping("ads/{id}/comments")
```
/ads можно на уровень класса вынести

BizinMitya commented on src/main/java/com/example/graduate_work_team2/controller/CommentController.java:
```diff
@@ -107,21 +111,27 @@ public ResponseEntity<HttpStatus> deleteComment(@PathVariable Integer adsId, @Pa
         return ResponseEntity.ok(updatedCommentDto);
     }
+
+    @GetMapping("/comments/{id}/image")
```
а это зачем? у комментария разве есть картинки?

dmitry-bizin commented on src/main/java/com/example/graduate_work_team2/mapper/impl/UserMapperImpl.java:
```diff
@@ -15,6 +15,7 @@ public class UserMapperImpl implements UserMapper {
 
     @Override
     public UserDto toDto(User entity) {
+        System.out.println("Был вызван метод маппера из User entity в UserDto"  );
```
а ниже не то же самое?)

dmitry-bizin commented on src/main/java/com/example/graduate_work_team2/service/impl/CommentServiceImpl.java:
```diff
@@ -0,0 +1,107 @@
+        Optional<User> user = userService.findAuthorizationUser();
+//        User simpleUser = user.get();
+//        String correctUser = simpleUser.getUsername();
+        if (comment.getAuthor().getEmail().equals(user.get().getEmail()) ) {
```
а проверка на админа?

dmitry-bizin commented on src/main/java/com/example/graduate_work_team2/service/impl/AdsServiceImpl.java:
```diff
@@ -0,0 +1,122 @@
+        newAds.setAuthor(userService.findAuthorizationUser().orElseThrow(UserNotFoundException::new));
+        Image newImage = imageService.uploadImage(imageFiles);
+        newAds.setImage(newImage);
+        adsRepository.save(newAds);
```
вот здесь надо написать `newAds = adsRepository.save(newAds);`, чтобы в newAds получить присвоенный БД id и чтобы этот id в маппере в ссылку прописался, из-за этого не показывается при создании объявления картинка

ITevgeniydav2022 committed:
Внесены изменения согласно замечаниям.
